### ISSUE TITLE
Altloc ID handling does not work properly

### ISSUE THREAD

**Issue by pschmidtke** [biotite-dev/biotite#194]
Hi I just tried to figure out how small molecules are currently handled within the structure objects (AtomArrayStack). 

If I extract for instance all residues using 
`residues: tuple=biotite.structure.get_residues(mystructure)`

then I get a nice tuple containing water, ions and amino-acids, but I don't see the small molecule in this list for instance on structure 5po6 (residue name 8SS).

How can I access these?

**Comment by danijoo** [comment]
Hi. Seems like you probably found a bug in the pdb parser. Some of the atoms are not present in the read in AtomArray:

```python
import biotite.structure.io as io
x = io.load_structure("5po6.pdb")
print(len(x))  # 2230

print(x[np.invert(struct.filter_amino_acids(x)) & (x.res_name != "HOH")])
> HET B     201  NA  NA     NA      -19.109    7.544    3.438
> HET B     202  EDO C1     C       -27.502    3.290    6.779
> HET B     202  EDO O1     O       -27.784    3.170    5.384
> HET B     202  EDO C2     C       -28.793    3.221    7.571
> HET B     202  EDO O2     O       -28.474    2.874    8.927
```

What we can see here is that `8SS` us apparently not included in the AtomArray at all. Also, the absolute number of atoms in the Array (2230) does not match the pdb, which contains a total of 2360 atoms.

``` bash
cat 5po6.pdb | grep ATOM | grep -v REMARK | wc -l
> 2068
cat 5po6.pdb | grep HETATM | grep -v REMARK | wc -l
> 292
```

On a side node, the error seems to be exclusive for reading pdbs, and does not occur for gro files:

```
x = io.load_structure("5po6.gro")
print(len(x))  # 2360
print(x[np.invert(struct.filter_amino_acids(x)) & (x.res_name != "HOH")])
          201  EDO C1     C        -9.190   12.500    1.260                
          201  EDO O1     O        -9.590   11.120    1.220                
          201  EDO C2     C        -9.430   13.120   -0.100                
          201  EDO O2     O       -10.830   13.070   -0.390                
          202  EDO C1     C        -7.680    8.850    0.220                
          202  EDO O1     O        -7.630   10.170   -0.320                
          202  EDO C2     C        -6.760    8.780    1.430                
          202  EDO O2     O        -7.480    8.110    2.480                
          203  8SS N1     N        -9.480    9.560    3.530                
          203  8SS N3     N        -7.020    9.310    1.030                
          203  8SS C4     C        -7.990   12.840   -1.220                
          203  8SS C5     C        -7.970   11.830   -0.290                
          203  8SS C6     C        -8.930   11.790    0.720                
          203  8SS C7     C        -8.910   10.720    1.740                
          203  8SS C8     C        -9.800   10.630    2.820                
          203  8SS C1     C        -9.900   12.790    0.760                
          203  8SS C2     C        -9.910   13.790   -0.190                
          203  8SS C3     C        -8.960   13.820   -1.180                
          203  8SS N2     N        -8.420    8.950    2.940                
          203  8SS C9     C        -8.050    9.640    1.850                
          203  8SS BR1    B        -8.940   15.170   -2.510                
          201  NA  NA     N       -19.110    7.540    3.440                
          202  EDO C1     C       -27.500    3.290    6.780                
          202  EDO O1     O       -27.780    3.170    5.380                
          202  EDO C2     C       -28.790    3.220    7.570                
          202  EDO O2     O       -28.470    2.870    8.930                
          203  EDO C1     C       -15.460    0.760   25.950                
          203  EDO O1     O       -14.700   -0.390   25.580                
          203  EDO C2     C       -14.560    1.990   26.040                
          203  EDO O2     O       -13.280    1.600   26.550                
```

**Comment by pschmidtke** [comment]
I also have the same bug for cif files

**Comment by pschmidtke** [comment]
![image](https://user-images.githubusercontent.com/15227239/80892390-f2eaa300-8cc9-11ea-9e3d-504cfdbebc33.png)
Why are you filtering before returning the structure in get_structure?

**Comment by padix-key** [comment]
I inspected the PDB file: The final filtering is actually the place where `8SS` is removed. Biotite only allows one altoc ID per residue, to avoid duplicate residues/atoms. However, by default it uses residues with altloc ID `A`. In this case `8SS` has the altoc ID `D`, which is filtered out.

Usually, alternative altoc IDs can be selected via the `altloc` parameter, but the parameter currently raises an error:
```python
cif_file = pdbx.PDBxFile.read(rcsb.fetch("5po6", "cif"))
structure = pdbx.get_structure(cif_file, model=1, altloc=[("A", 203, "D")])
```
```
NameError: name 'altloc' is not defined
```

In summary, there are two issues to fix:
- By default, use the first altloc ID, instead of altloc ID `A`
- Fix `NameError` in `filter_altloc()`

**Comment by padix-key** [comment]
If you need a fix right now, please use the snippet from above and change the line
```python
chain, residue, altoc = loc
```
to
```python
chain, residue, altloc = loc
```
in `biotite/structure/filter.py`.

I will prepare a PR for this issue in the next few days.

**Comment by pschmidtke** [comment]
How is this affecting antibody structures? Especially the residue unicity?
For instance in structure 1dba you have 2 residues with the same residue id but an insertion code (that's a crappy pdb convention as well): 

![image](https://user-images.githubusercontent.com/15227239/80912814-0c2e3680-8d40-11ea-8f56-9db044399b92.png)

**Comment by pschmidtke** [comment]
I think it would be easier to return the full structure (without filter) and let the user filter the way he wants - so filtering could be made optional when you load a structure via a function argument?

**Comment by padix-key** [comment]
The filter does not affect insertion codes, so these should be always included. In contrast, altlocs mean that the same atom(s) may be present at two or more different positions. Hence, I think it is not meaningful to have the same atom twice (with different coordinates) in the same model. However, if there is a use case for it, we could add a parameter to `get_structure()` to omit filtering at all. Maybe, something like `altloc="all"`?

**Comment by pschmidtke** [comment]
Yes, but the filter you currently have in place works for residue (id or name?), chain and altloc. In the example of 1dba what happens with residue 52 there? You have an isoleucine and asparagine with the same description. From my understanding this will be filtered out as well? will have to test ;)

**Comment by padix-key** [comment]
Indeed this is an issue. If you select an altloc, this would select both residues, the `ASN` and `ILE`, with the same altloc. Currently, there is no way to select residue `52` with e.g. altloc `A` and residue `52A` with e.g. altloc `B`.

In summary, the following changes need to be implemented:

- By default, use the first altloc ID, instead of altloc ID A
- Fix NameError in filter_altloc()
- Add way to select separate altloc for residues with insertion codes
- Add parameter to omit filtering at all

**Comment by pschmidtke** [comment]
In 1dba these are not altlocs but insertion codes. And the expected behaviour is to read both residues as they are really both there one after the other, they are just weirdly numbered.



________________________________

**Comment by danijoo** [comment]
Altloc and insertion handling is really a cumberstone in PDB formats since they are hard to read and interpret in general, mainly because they are sometimes abused to show something they are not intended for. I think having an `altloc="all"` flag that defaults to `altloc=None` would be a good way to handle this.

**Comment by padix-key** [comment]
> In 1dba these are not altlocs but insertion codes. And the expected behaviour is to read both residues as they are really both there one after the other, they are just weirdly numbered.

As these residues have no altloc, both are included in the resulting `AtomArray`, one with `''`, the other one with `'A'` in the `ins_code` annotation array.
I meant that if these residues had altlocs, there is currently no way to select two different altlocs for these residues.

**Comment by padix-key** [comment]
The `ParmEd` package uses a string parameter to select altlocs:
https://parmed.github.io/ParmEd/html/api/parmed/parmed.formats.pdb.html#parmed.formats.pdb.CIFFile.write.

- `'all'`: return all alternate locations
- `'first'` : return only the first alternate locations
- `'occupancy'` : return the one with the largest occupancy. If two conformers have the same occupancy, the first one to occur is printed.

`alt_loc` would be another optional annotation array, that is added to the `AtomArray` or `AtomArrayStack`, if `'all'` is chosen. Thus the user can select the altloc they want in the same way they would select atoms by any other criteria.

In case of *Biotite* `'first'` would be default, in consistency with the current behavior. If the user would like to select.

Do you think this is a sensible solution to this issue?

**Comment by padix-key** [comment]
I will merge this PR then.
